packages/react/src/editor/BlockNoteView.tsx

```diff
@@ -77,32 +41,55 @@ export function BlockNoteView<
     children?: ReactNode;
   } & HTMLAttributes<HTMLDivElement>
 ) {
-  const {
-    theme = { light: lightDefaultTheme, dark: darkDefaultTheme },
-    ...rest
-  } = props;
+  const { editor, className, theme, children, ...rest } = props;
 
-  const preferredTheme = usePrefersColorScheme();
+  const systemColorScheme = usePrefersColorScheme();
 
-  useEffect(() => {
+  const editorColorScheme = useMemo(() => {
     if (theme === "light") {
-      themeToCSSVariables(lightDefaultTheme);
-    } else if (theme === "dark") {
-      themeToCSSVariables(darkDefaultTheme);
-    } else if (
-      typeof theme === "object" &&
-      "light" in theme &&
-      "dark" in theme
-    ) {
-      themeToCSSVariables(theme[preferredTheme === "dark" ? "dark" : "light"]);
-    } else {
-      themeToCSSVariables(theme);
+      return "light";
+    }
+
+    if (theme === "dark") {
+      return "dark";
     }
-  }, [preferredTheme, theme]);
+
+    if (typeof theme === "object" && "light" in theme && "dark" in theme) {
+      if ("light" in theme && "dark" in theme) {
+        applyCSSVariablesFromTheme(
+          theme[systemColorScheme === "dark" ? "dark" : "light"],
+          editor.domElement.parentElement!
+        );
+        return systemColorScheme === "dark" ? "dark" : "light";
+      }
+
+      applyCSSVariablesFromTheme(theme, editor.domElement.parentElement!);
+      return undefined;
+    }
+
+    return systemColorScheme === "dark" ? "dark" : "light";
+  }, [systemColorScheme, editor.domElement, theme]);
```

Comment:
Did you test changing themes, including from a custom theme to a built in one, etc? I doubt this works as the css variables aren't removed